langchain/agents/agent.py

```diff
@@ -762,14 +762,25 @@ def _take_next_step(
                 **inputs,
             )
         except Exception as e:
-            if not self.handle_parsing_errors:
+            if isinstance(self.handle_parsing_errors, bool):
+                raise_error = not self.handle_parsing_errors
+            else:
+                raise_error = False
+            if raise_error:
                 raise e
-            text = str(e).split("`")[1]
-            observation = "Invalid or incomplete response"
+            text = str(e)
+            if isinstance(self.handle_parsing_errors, bool):
+                observation = "Invalid or incomplete response"
+            elif isinstance(self.handle_parsing_errors, str):
+                observation = self.handle_parsing_errors
+            elif callable(self.handle_parsing_errors):
+                observation = self.handle_parsing_errors(text)
```

Comment:
I wonder if it's more flexible to pass in the original exception, or just the string form of it?